src/almo_scf_optimizer.F

```diff
@@ -7605,5 +7427,2016 @@ SUBROUTINE hessian_diag_apply(matrix_grad, matrix_step, matrix_S_ao, &
+
+      REAL(KIND=dp), INTENT(INOUT)                       :: step_size_out
+      TYPE(dbcsr_type), DIMENSION(:), INTENT(IN)         :: metric_in, position_in, direction_in
+      REAL(KIND=dp), INTENT(IN)                          :: trust_radius_in
+      TYPE(dbcsr_type), DIMENSION(:), INTENT(IN)         :: quench_t_in
+      REAL(KIND=dp), INTENT(IN)                          :: eps_filter_in
+
+      INTEGER                                            :: isol, ispin, nsolutions, &
+                                                            nsolutions_found, nspins
+      INTEGER, ALLOCATABLE, DIMENSION(:)                 :: nocc
+      REAL(KIND=dp)                                      :: discrim_sign, discriminant, solution, &
+                                                            spin_factor, temp_real
+      REAL(KIND=dp), DIMENSION(3)                        :: coef
+      TYPE(dbcsr_type), ALLOCATABLE, DIMENSION(:)        :: m_temp_no
+
+      step_size_out = 0.0_dp
+
+      nspins = SIZE(position_in)
+      IF (nspins == 1) THEN
+         spin_factor = 2.0_dp
+      ELSE
+         spin_factor = 1.0_dp
+      ENDIF
+
+      ALLOCATE (nocc(nspins))
+      ALLOCATE (m_temp_no(nspins))
+
+      coef(:) = 0.0_dp
+      DO ispin = 1, nspins
+
+         CALL dbcsr_create(m_temp_no(ispin), &
+                           template=direction_in(ispin))
+
+         CALL dbcsr_get_info(direction_in(ispin), &
+                             nfullcols_total=nocc(ispin))
+
+         CALL dbcsr_copy(m_temp_no(ispin), quench_t_in(ispin))
+         CALL dbcsr_multiply("N", "N", 1.0_dp, &
+                             metric_in(1), &
+                             position_in(ispin), &
+                             0.0_dp, m_temp_no(ispin), &
+                             retain_sparsity=.TRUE.)
+         CALL dbcsr_filter(m_temp_no(ispin), eps_filter_in)
+         CALL dbcsr_dot(position_in(ispin), m_temp_no(ispin), temp_real)
+         coef(3) = coef(3) + temp_real/nocc(ispin)
+         CALL dbcsr_dot(direction_in(ispin), m_temp_no(ispin), temp_real)
+         coef(2) = coef(2) + 2.0_dp*temp_real/nocc(ispin)
+         CALL dbcsr_copy(m_temp_no(ispin), quench_t_in(ispin))
+         CALL dbcsr_multiply("N", "N", 1.0_dp, &
+                             metric_in(1), &
+                             direction_in(ispin), &
+                             0.0_dp, m_temp_no(ispin), &
+                             retain_sparsity=.TRUE.)
+         CALL dbcsr_filter(m_temp_no(ispin), eps_filter_in)
+         CALL dbcsr_dot(direction_in(ispin), m_temp_no(ispin), temp_real)
+         coef(1) = coef(1) + temp_real/nocc(ispin)
+
+         CALL dbcsr_release(m_temp_no(ispin))
+
+      ENDDO !ispin
+
+      DEALLOCATE (nocc)
+      DEALLOCATE (m_temp_no)
+
+      coef(:) = coef(:)*spin_factor
+      coef(3) = coef(3) - trust_radius_in*trust_radius_in
+
+      ! solve the quadratic equation
+      discrim
```

Comment:
might make sense to turn this subroutine into a function, but I guess that's a matter of style

Reply:
This subroutine will be modified in the future to return a matrix. That is why it is not implemented as a function.